library/common/include/c_types.h

```diff
@@ -0,0 +1,130 @@
+#pragma once
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+// NOLINT(namespace-envoy)
+
+/**
+ * Handle to an Envoy engine instance. Valid only for the lifetime of the engine and not intended
+ * for any external interpretation or use.
+ */
+typedef uint64_t envoy_engine_t;
+
+/**
+ * Handle to an outstanding Envoy HTTP stream. Valid only for the duration of the stream and not
+ * intended for any external interpretation or use.
+ */
+typedef uint64_t envoy_stream_t;
+
+/**
+ * Result codes returned by all calls made to this interface.
+ */
+typedef enum { ENVOY_SUCCESS, ENVOY_FAILURE } envoy_status_t;
+
+/**
+ * Error code associated with terminal status of a HTTP stream.
+ */
+typedef enum { ENVOY_STREAM_RESET } envoy_error_code_t;
+
+/**
+ * Common abstraction for strings.
+ */
+typedef struct {
+  uint64_t length;
+  const char* data;
+} envoy_string;
+
+/**
+ * Holds raw binary data as an array of bytes.
+ */
+typedef struct {
+  uint64_t length;
+  uint8_t* bytes;
+} envoy_data;
+
+/**
+ * Holds data about an HTTP stream.
+ */
+typedef struct {
+  // Status of the Envoy HTTP stream. Note that the stream might have failed inline.
+  // Thus the status should be checked before pursuing other operations on the stream.
+  envoy_status_t status;
+  // Handle to the Envoy HTTP Stream.
+  envoy_stream_t stream;
+} envoy_stream;
+
+/**
+ * Holds a single key/value header.
+ */
+typedef struct {
+  envoy_string key;
+  // Multiple header values for the same header key are supported via a comma-delimited string.
+  envoy_string value;
+} envoy_header;
+
+/**
+ * Holds an HTTP header map as an array of envoy_header structs.
+ */
+typedef struct {
+  // Number of header elements in the array.
+  uint64_t length;
+  // Array of headers.
+  envoy_header* headers;
+} envoy_headers;
+
+// Convenience constant to pass to function calls with no data.
+// For example when sending a headers-only request.
+const envoy_data envoy_nodata = {0, NULL};
+
+/**
+ * Error struct.
+ */
+typedef struct {
+  envoy_error_code_t error_code;
+  envoy_string message;
+} envoy_error;
+
+#ifdef __cplusplus
+extern "C" { // function pointers
+#endif
+/**
+ * Called when all headers get received on the async HTTP stream.
+ * @param headers, the headers received.
+ * @param end_stream, whether the response is headers-only.
+ */
+typedef void (*on_headers)(envoy_headers headers, bool end_stream, void* context);
```

Comment:
Do we also need an observer for `on_metadata`? Looks like that's missing here

Reply:
Envoy doesn't actually support it yet, but this is still a good callout. Let's add it so we can have everything rigged for when support does land.